text/0000-alloca.md

```diff
@@ -0,0 +1,147 @@
+
+Apart from that, this works much like an unboxed array, except the size is
+determined at runtime. Since the memory resides on the stack, be careful
+not to exceed the stack limit (which depends on your operating system),
+otherwise the resulting stack overflow will at best kill your program. You
+have been warned.
+
+Valid uses for this is mostly within embedded system without heap allocation.
+```
+
+Also add an example (perhaps a sort algorithm that uses some scratch space that will be heap-allocacted with `std` and
+stack-allocated with `#[no_std]` (noting that the function would not be available on no-std systems at all were it not
+for this feature).
+
+Do not `pub use` it from `std::mem` to drive the point home.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Even more stack usage means the dreaded stack limit will probably be reached even sooner. Overflowing the stack space
+leads to segfaults at best and undefined behavior at worst. On unices, the stack can usually be extended at runtime,
+whereas on Windows stack size is set at link time (default to 1MB).
+
+- Adding this will increase implementation complexity and require support from possible alternative implementations /
+backends (e.g. Cretonne, WebASM).
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do nothing. Rust works well without it (there's the issue mentioned in the "Motivation" section though). `SmallVec`s
+work well enough and have the added benefit of limiting stack usage.
+
+-  `mem::with_alloc<T, F: Fn([T]) -> U>(elems: usize, code: F) -> U` This has the benefit of reducing API surface, and
+introducing rightwards drift, which makes it more unlikely to be used too much. However, it also needs to be
+monomorphized for each function (instead of only for each target type), which will increase compile times.
+
+- dynamically sized arrays are a potential solution, however, those would need to have a numerical type that is only
+fully known at runtime, requiring complex type system gymnastics.
+
+- use a macro instead of a function (analogous to `print!(..)`), which could insert the LLVM alloca builtin.
+
+- mark the function as `unsafe` due to the potential stack overflowing problem.
+
+- Copy the design from C `fn alloca()`, possibly wrapping it later.
+
+- Use escape analysis to determine which allocations could be moved to the stack. This could potentially benefit even
+more programs, because they would benefit from increased allocation speed without the need for change. The deal-breaker
+here is that we would also lose control to avoid the listed drawback, making programs crash without recourse. Also the
+compiler would become somewhat more complex (though a simple incomplete escape analysis implementation is already in
+[clippy](https://github.com/Manishearth/rust-clippy).
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Could we return the slice directly (reducing visible complexity)?
```

Comment:
I may have missed it, but I'm not sure I understand why we wouldn't be able to just return the slice?